metrics/scaling/k8s_scale.sh

```diff
@@ -0,0 +1,262 @@
+# $2 is the number of pod/containers under test
+grab_stats() {
+	local launch_time_ms=$1
+	local n_pods=$2
+	info "And grab some stats"
+	# Tell mpstat to measure over a short period, not only so we get slightly de-noised data, but also
+	# if you don't tell it the period, you will get the avg since boot, which is not what we want.
+	local cpu_idle=$(kubectl exec -ti ${stats_pod} -- sh -c "mpstat -u 3 1 | tail -1 | awk '{print \$11}'" | sed 's/\r//')
+	local mem_free=$(kubectl exec -ti ${stats_pod} -- sh -c "free | tail -2 | head -1 | awk '{print \$4}'" | sed 's/\r//')
+
+	info "idle [$cpu_idle] free [$mem_free] launch [$launch_time]"
+
+	# Annoyingly, it seems sometimes once in a while we don't get an answer!
+	# We should really retry, but for now, make the json valid at least
+	cpu_idle=${cpu_idle:-0}
+	mem_free=${mem_free:-0}
+
+	local json="$(cat << EOF
+	{
+		"n_pods": {
+			"Result": ${n_pods},
+			"Units" : "int"
+		},
+		"cpu_idle": {
+			"Result": ${cpu_idle},
+			"Units" : "%"
+		},
+		"launch_time": {
+			"Result": $launch_time_ms,
+			"Units" : "ms"
+		},
+		"mem_free": {
+			"Result": ${mem_free},
+			"Units" : "kb"
+		}
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+}
+
+init() {
+	info "Initialising"
+	info "Checking k8s accessible"
+	local worked=$( kubectl get nodes > /dev/null 2>&1 && echo $? || echo $? )
+	if [ "$worked" != 0 ]; then
+		die "kubectl failed to get nodes"
+	fi
+
+	info $(get_num_nodes) "k8s nodes found"
+	# We could check we have just the one node here - right now this is a single node
+	# test!! - because, our stats gathering is rudimentry, as k8s does not provide
+	# a nice way to do it (unless you want to parse 'descibe nodes')
+	# Have a read of https://github.com/kubernetes/kubernetes/issues/25353
+
+	# FIXME - check the node(s) can run enough pods - check 'max-pods' in the
+	# kubelet config - from 'kubectl describe node -o json' ?
+
+	# Launch our stats gathering pod
+	kubectl apply -f ${SCRIPT_PATH}/${stats_pod}.yaml
+	kubectl wait --for=condition=Ready pod "${stats_pod}"
+
+	# FIXME - we should probably 'warm up' the cluster with the container image(s) we will
+	# use for testing, otherwise the download time will likely be included in the first pod
+	# boot time.
+
+	# And now we can set up our results storage then...
+	metrics_json_init "k8s"
+	save_config
+}
+
+save_config(){
+	metrics_json_start_array
+
+	local json="$(cat << EOF
+	{
+		"testname": "${TEST_NAME}",
+		"NUM_PODS": "${NUM_PODS}",
+		"STEP": "${STEP}",
+		"wait_time": "${wait_time}",
+		"delete_wait_time": "${delete_wait_time}",
+		"settle_time": "${settle_time}"
+	}
+EOF
+)"
+	metrics_json_add_array_element "$json"
+	metrics_json_end_array "Config"
+}
+
+run() {
+	info "Running test"
+
+	trap cleanup EXIT QUIT KILL
+
+	metrics_json_start_array
+	for reqs in $(seq ${STEP} ${STEP} ${NUM_PODS}); do
+		info "Testing replicas ${reqs} of ${NUM_PODS}"
+		# Generate the next yaml file
+
+	local runtime_command
```

Comment:
just noticed, I think indentation is bust in this section - it is all in the `for` loop, so some of the lines need to be pushed in one tab.

Reply:
we should incorporate these, maybe as git pre-commit hooks?
https://github.com/mvdan/sh
https://github.com/koalaman/shellcheck

Reply:
@krsna1729 eek. maybe.... I think we opted out of `shellcheck` over in kata land as there were some things it moaned about that were particularly hard to get around iirc...

Let's see if anybody finds the time to try these out on the code.... and, reserve the right to not inflict them if they pain barrier will be too high ;-)